[MRG] Linear One-Class SVM using SGD implementation

albertcthomas opened the pull request:
#### What does this implement/fix? Explain your changes.
This implements a linear version of the One-Class SVM based on the SGD implementation. This implementation thus scales linearly with the number of samples and has a `partial_fit` method. Combining this implementation with kernel approximation techniques, we can approximate the solution of a kernelized `OneClassSVM` and still benefit from the training time complexity improvement (see example and benchmark below).

The optimization problem of the One-Class SVM can be written as an optimization problem that is very close to the ones solved by the SGD implementation (see the [doc for details](https://14573-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/sgd.html#sgd-online-one-class-svm)). This implementation thus requires very few changes in the SGD cython code.

Benchmark comparing `OneClassSVM` and `SGDOneClassSVM` in terms of training time and AUC (n: number of training samples, d: number of features). The training size has been reduced for some of the datasets for LibSVM to finish in a decent time.

![train_time](https://user-images.githubusercontent.com/15966638/32114362-9ce36406-bb43-11e7-8ada-8f0a304ad334.png)

![auc](https://user-images.githubusercontent.com/15966638/32114366-a2ac1522-bb43-11e7-8ca3-e0206c82f9ef.png)

Toy example
![toy](https://user-images.githubusercontent.com/15966638/32114391-b086bd64-bb43-11e7-9da1-6de1540cc41d.png)

#### Any other comments?
This is still WIP because the tests can be refactored. Any comment is more than welcome.
cc @agramfort

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

albertcthomas committed:
sgd for one class svm

amueller commented:
This looks cool, thank you.
Needs a user guide and maybe adding to some of the outlier detection examples? I think the docs should also be pretty explicit that this mostly makes sense with some kernel approximation in front of it (right?).

Since this is a pretty significant contribution, expect that this will take some time to review...

albertcthomas commented:
> I think the docs should also be pretty explicit that this mostly makes sense with some kernel approximation in front of it (right?).

Right

albertcthomas commented:
For the part of the doc explaining how the SGD implementation is used see [here](https://14573-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/sgd.html#sgd-online-one-class-svm).
I will add this algorithm to the outlier detection example suggested in #10004 once it is merged.

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+        `partial_fit`.
+        Defaults to 5. Defaults to 1000 from 0.21, or if tol is not None.
+
+        .. versionadded:: 0.19
```
I would remove this. `SGDOneClassSVM` is to be added in v0.20 at best

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+        when (loss > previous_loss - tol). Defaults to None.
+        Defaults to 1e-3 from 0.21.
+
+        .. versionadded:: 0.19
```
I would remove this

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+
+class SGDOneClassSVM(BaseSGD):
+    """Solves linear One-Class SVM using Stochastic Gradient Descent.
+
```
Maybe add:
```
    .. versionadded:: 0.20
```

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+        The maximum number of passes over the training data (aka epochs).
+        It only impacts the behavior in the ``fit`` method, and not the
+        `partial_fit`.
+        Defaults to 5. Defaults to 1000 from 0.21, or if tol is not None.
```
I guess as this class is new, we can directly use the future default value.
what do you think?

albertcthomas replied:
done

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+    tol : float or None, optional
+        The stopping criterion. If it is not None, the iterations will stop
+        when (loss > previous_loss - tol). Defaults to None.
+        Defaults to 1e-3 from 0.21.
```
Same here, I would use directly future default value.

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+                 warm_start=False, average=False):
+
+        if not(0 < nu <= 1):
+            raise ValueError("nu must be in (0, 1]")
```
We usually avoid checking  parameters in the `__init__`, since if the value is changed with `set_params`, it has to be checked again during the fit.

albertcthomas replied:
I agree but there is a `self.validate_params` in `BaseSGD.__init__`. If I remove the check about `nu` in `SGDOneClassSVM.__init__` no error message or a non-informative error message is raised concerning the value of `alpha` which is not a parameter of the `SGDOneClassSVM`

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+    offset_ : array, shape (1,)
+        Offset used to define the decision function from the raw scores.
+        We have the relation: decision_function = score_samples - offset_.
+        The offset is the opposite of intercept_ and is provided for
```
Is it `offset_ = 1 - intercept_` or `offset_ = - intercept_`?

albertcthomas replied:
I removed this sentence as intercept_ is not an attribute of the estimator, only offset_ is.

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+        """
+
+        if not(0 < self.nu <= 1):
+            raise ValueError("nu must be in (0, 1]")
```
Can you update the message with the incorrect `nu` value?

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,416 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+                                 offset_decay=offset_decay)
+
+    def _fit(self, X, alpha, C, loss, learning_rate, coef_init=None,
+             offset_init=None, sample_weight=None, one_class=0,
```
`one_class` is not used

albertcthomas replied:
thanks

TomDLT commented on benchmarks/bench_online_ocsvm.py:
```diff
@@ -0,0 +1,290 @@
+
+        print('----------- Online OCSVM ------------')
+        nystroem = Nystroem(gamma=gamma, random_state=random_state)
+        online_ocsvm = SGDOneClassSVM(nu=nu, random_state=random_state)
```
You need to specify `tol` or `max_iter` to be future-proof and avoid a `FutureWarning`

TomDLT commented on sklearn/linear_model/base.py:
```diff
@@ -49,7 +49,8 @@ def make_dataset(X, y, sample_weight, random_state=None):
 
     This also returns the ``intercept_decay`` which is different
-    for sparse datasets.
+    for sparse datasets. ``intercept_decay`` is not used for
```
Why? Wouldn't we have the same offset oscillation issues when the data is sparse?

Also, I am not sure this is right place to put this comment. I would move it as a comment before the call to `make_dataset` in `SGDOneClassSVM._fit`

albertcthomas replied:
I moved the comment to `_fit_oneclass` before the call to `make_dataset`.

For the moment I have preferred to allow the user to set this value in the `fit` method using the `offset_decay` parameter and not use 0.01 by default for sparse data (and 1 for dense data). The default value of the `offset_decay` is set to 1  and this is the value used in the benchmark and in the example but more experiments would be needed.

Besides I don't really understand this condition for sparse data only. I understand that we might want to have a different learning rate for the intercept as explained in the Stochastic Gradient Descent Tricks by Leon Bottou and as done in the code available at [his sgd page](http://leon.bottou.org/projects/sgd). However it seems that enforcing a smaller learning rate for the intercept is done for any kind of data and not only for sparse data:

````cpp
/// Perform one iteration of the SGD algorithm with specified gains
void
SvmSgd::trainOne(const SVector &x, double y, double eta)
{
  double s = dot(w,x) / wDivisor + wBias;
  // update for regularization term
  wDivisor = wDivisor / (1 - eta * lambda);
  if (wDivisor > 1e5) renorm();
  // update for loss term
  double d = LOSS::dloss(s, y);
  if (d != 0)
    w.add(x, eta * d * wDivisor);
  // same for the bias
#if BIAS
  double etab = eta * 0.01;
#if REGULARIZED_BIAS
  wBias *= (1 - etab * lambda);
#endif
  wBias += etab * d;
#endif
}
````

albertcthomas commented on benchmarks/bench_online_ocsvm.py:
```diff
@@ -0,0 +1,290 @@
+    auc_libsvm = auc(x_axis, tpr_libsvm)
+
+    results_libsvm[dat] = ([fit_time_libsvm, predict_time_libsvm,
+                            auc_libsvm, int(n_samples_train),
```
I removed it

albertcthomas commented:
Thanks for the review @TomDLT! The estimator is now using the future default values of `max_iter` and `tol`. For the moment, `offset_decay` is a parameter of the `fit` method but we should maybe remove it depending on the benchmark results. I do not know yet if we should consider the same decay as the other SGD estimators for sparse data (see my [comment above](https://github.com/scikit-learn/scikit-learn/pull/10027#discussion_r150195106))

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1341,3 +1354,412 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+                 learning_rate="optimal", eta0=0.0, power_t=0.5,
+                 warm_start=False, average=False):
+
+        if not(0 < nu <= 1):
```
What about merging all the `nu` checks into `_validate_params`?
```py
    def _validate_params(self, set_max_iter=True):
        """Validate input params. """
        nu = self.nu
        if not(0 < self.nu <= 1):
            raise ValueError(
		"self.nu must be in (0, 1], got nu=%f" % self.nu)

        super(SGDOneClassSVM, self)._validate_params(
	    set_max_iter=set_max_iter)
```

TomDLT reviewed (commented):
The idea behind having a smaller intercept update for sparse data is that the intercept is updated *at each sample*, whereas a particular feature coefficient is updated *only when this feature is nonzero in the current sample*. Thus the intercept is updated much more often that other features, hence the decay to balance this effect.

I don't mind having a parameter `offset_decay`, but I would be in favor of homogeneity among SGD classes, as it seems that `SGDOneClassSVM` is not fundamentally different. Also, I would put the parameter in  `__init__`, and not in `fit`.

jnothman commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-2294b84cc64f18116bacf106a4160e33b7fff097)

**new alerts:**

* 1 for Unreachable code

---

*Comment posted by [lgtm.com](https://lgtm.com)*

albertcthomas commented:
Thanks for the clarification @TomDLT. Let's use the same decay for `SGDOneClassSVM` then.

albertcthomas commented:
I removed the `offset_decay` parameter as I'm using the same default values as the other SGD estimators.

jnothman commented:
This pull request introduces 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-16ea80bce5d14452fe6fcc46b0ebd76de32005cb)

**new alerts:**

* 1 for Unreachable code
* 1 for __init__ method calls overridden method

---

*Comment posted by [lgtm.com](https://lgtm.com)*

albertcthomas commented:
I rebased on master and added the algorithm to the example `plot_anomaly_comparison.py`
![anomaly_comparison](https://user-images.githubusercontent.com/15966638/48867722-5e42fc80-edd7-11e8-8152-9699b8616be1.png)

albertcthomas commented:
I also fixed the benchmark so that the shuttle dataset is now downloaded with `fetch_openml`. Results are:
Training time:
![training-time](https://user-images.githubusercontent.com/15966638/48867766-9a765d00-edd7-11e8-87b2-e44bc297f0c1.png)
AUC:
![auc](https://user-images.githubusercontent.com/15966638/48867780-a7934c00-edd7-11e8-9c14-f8ef4ca7b1fa.png)

TomDLT reviewed (commented):
Some minor remarks on sphinx link rendering.

Common tests are also failing.

albertcthomas commented:
Thanks again for the review @TomDLT! I will double check the doc when the rendering is available.

albertcthomas commented on doc/modules/sgd.rst:
```diff
@@ -204,6 +204,55 @@ For regression with a squared loss and a l2 penalty, another variant of
+approximation techniques, :class:`SGDOneClassSVM` can be used to approximate
+the solution of a kernelized :class:`sklearn.svm.OneClassSVM` with a linear
+complexity in the number of samples. Note that the complexity of a kernelized
+:class:`sklearn.svm.OneClassSVM` is at best quadratic in the number of samples.
```
Actually, removing `sklearn` seems to break the link (see [CircleCI](https://38928-843222-gh.circle-artifacts.com/0/doc/modules/sgd.html#sgd-online-one-class-svm))

albertcthomas commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1585,3 +1598,437 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+
+    See also
+    --------
+    OneClassSVM
```
I need to put `sklearn.svm.OneClassSVM`

albertcthomas commented on doc/modules/outlier_detection.rst:
```diff
@@ -169,6 +169,17 @@ but regular, observation outside the frontier.
+approximate the solution of a kernelized :class:`svm.OneClassSVM` whose
+complexity is at best quadratic in the number of samples. See section
+:ref:`sgd_online_one_class_svm` for more details.
+
```
I will fix the link of the example, I forgot to add linear_model...

albertcthomas commented:
I will have to investigate the last failing test... apparently clf.coef_ is reallocated when doing a partial_fit. But this fails only on python2,7 for appveyor.

albertcthomas commented:
The failing test is not failing anymore so this appears to be random.

jnothman commented:
Is this still of interest @albertcthomas? Lots of red crosses.

albertcthomas commented:
There should be less red crosses now that I rebased. This PR was almost already done except for the tests. Should be good now.

albertcthomas commented:
Rendered doc: [outlier detection](https://49192-843222-gh.circle-artifacts.com/0/doc/modules/outlier_detection.html#scaling-up-the-one-class-svm), [sgd](https://49192-843222-gh.circle-artifacts.com/0/doc/modules/sgd.html#online-one-class-svm) and [user guide](https://49192-843222-gh.circle-artifacts.com/0/doc/modules/generated/sklearn.linear_model.SGDOneClassSVM.html).

bthirion commented on benchmarks/bench_online_ocsvm.py:
```diff
@@ -0,0 +1,290 @@
+n_axis = 1000
+x_axis = np.linspace(0, 1, n_axis)
+
+# datasets available: ['http', 'smtp', 'SA', 'SF', 'shuttle', 'forestcover']
```
to be removed

bthirion commented on examples/plot_anomaly_comparison.py:
```diff
@@ -87,6 +100,14 @@
                                          random_state=42)),
     ("Local Outlier Factor", LocalOutlierFactor(
         n_neighbors=35, contamination=outliers_fraction))]
+# SGDOneClassSVM must be used with a kernel approximation to give similar
```
I'd put it right after OCSVM.

bthirion commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1593,3 +1606,440 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+            eta = eta0
+        'optimal': [default]
+            eta = 1.0 / (alpha * (t + t0))
+            where t0 is chosen by a heuristic proposed by Leon Bottou.
```
Give a ref ?

albertcthomas replied:
This is the same docstring as `SGDClassifier` and `SGDRegressor`

bthirion commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1593,3 +1606,440 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+            set_max_iter=set_max_iter,
+            for_partial_fit=for_partial_fit)
+
+    def _fit_oneclass(self, X, alpha, C, sample_weight,
```
'_fit_one_class' would be better

bthirion commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1593,3 +1606,440 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+            self.coef_ = None
+            self.offset_ = None
+
+        if self.average > 0:
```
`if self.average:` should be enough, no ?

bthirion reviewed (commented):
Looks great overall; only minor syntactic comments.

albertcthomas commented:
Thanks for the review @bthirion

banilo commented on doc/modules/outlier_detection.rst:
```diff
@@ -106,11 +106,14 @@ detection.
-hereunder.
+does not perform very well for outlier detection.
+:class:`linear_model.SGDOneClassSVM` approximates the kernelized One-Class SVM
+(implemented in :class:`svm.OneClassSVM`) thanks to kernel approximation
```
pleonasm in this and previous line

banilo commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -1593,3 +1606,440 @@ def __init__(self, loss="squared_loss", penalty="l2", alpha=0.0001,
+            self.offset_ = 1 - intercept
+            self.offset_ = np.atleast_1d(self.offset_)
+            self.t_ += self.n_iter_ * n_samples
+
```
empty line

DanyYan commented:
Hello. I want to know which version of sklearn can i use the SGDOneClassSVM. I can't find it in the scikit-learn v0.21.3

albertcthomas commented:
Hi @DanyYan, this algorithm is still under review and not part of a scikit-learn official release. To use it you have to checkout this pull request locally. Any feedback is more than welcome :).

akshuga commented:
I cannot see [guide page](https://49192-843222-gh.circle-artifacts.com/0/doc/modules/generated/sklearn.linear_model.SGDOneClassSVM.html) now.
What should I refer for the usage?

albertcthomas commented:
@akshuga you can refer to the [example](https://github.com/scikit-learn/scikit-learn/blob/98daa22e29c6027f8ab39eae355eb9ec02c06fcd/examples/linear_model/plot_sgdocsvm_vs_ocsvm.py) created with this PR. If you like this feature don't hesitate to add a :+1: to the [original comment](https://github.com/scikit-learn/scikit-learn/pull/10027#issue-149232707) of this PR.

cmarmo commented:
Hello @albertcthomas, I'm reviving old PRs that need review: if you are still interested in working on this, do you mind fixing conflicts? Thanks for your work and your patience!

albertcthomas commented:
Sure I’ll do it as soon as possible.

On Tue 15 Dec 2020 at 13:19, Chiara Marmo <notifications@github.com> wrote:

> Hello @albertcthomas <https://github.com/albertcthomas>, I'm reviving old
> PRs that need review: if you are still interested in working on this, do
> you mind fixing conflicts? Thanks for your work and your patience!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10027#issuecomment-745252862>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZ2DLULS23EOYNK2B4NCF3SU5H3RANCNFSM4EBFHFZA>
> .
>

albertcthomas commented on benchmarks/bench_online_ocsvm.py:
```diff
@@ -0,0 +1,279 @@
+x_axis = np.linspace(0, 1, n_axis)
+
+datasets = ['http', 'smtp', 'SA', 'SF', 'forestcover']
+
```
I removed the shuttle dataset as it is not available anymore

albertcthomas commented:
Results of the benchmark
![Figure_1](https://user-images.githubusercontent.com/15966638/103945096-f83f9d80-5134-11eb-920e-596fd5471d39.png)
![Figure_2](https://user-images.githubusercontent.com/15966638/103945120-fd045180-5134-11eb-9b8e-0dfeee9e9903.png)
![Figure_3](https://user-images.githubusercontent.com/15966638/103945132-0097d880-5135-11eb-9af7-bc1606b26f31.png)

albertcthomas commented:
@cmarmo conflicts are fixed, I am not sure why the test is failing on py36_ubuntu_atlas_32bit but passing on all other platforms.

albertcthomas commented:
Hm actually it is not just one job that is failing. I'll have a look.

albertcthomas committed:
trying test with almost equal

albertcthomas commented:
All green now.

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Subset of the training data
```
```suggestion
            Subset of the training data.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training data
```
```suggestion
            Training data.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+            The initial coefficients to warm-start the optimization.
+
+        offset_init : array, shape (n_classes,)
+            The initial offset to warm-start the optimization
```
```suggestion
            The initial offset to warm-start the optimization.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
```
```suggestion
        X : {array-like, sparse matrix}, shape (n_samples, n_features)
            Testing data.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
```
```suggestion
        X : {array-like, sparse matrix}, shape (n_samples, n_features)
            Testing data.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+        return decisions.ravel()
+
+    def score_samples(self, X):
+        """Raw scoring function of the samples
```
```suggestion
        """Raw scoring function of the samples.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+        return score_samples
+
+    def predict(self, X):
+        """Return labels (1 inlier, -1 outlier) of the samples
```
```suggestion
        """Return labels (1 inlier, -1 outlier) of the samples.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
```
```suggestion
        X : {array-like, sparse matrix}, shape (n_samples, n_features)
            Testing data.
```

TomDLT commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1596,3 +1608,435 @@ def _more_tags(self):
+
+class SGDOneClassSVM(BaseSGD, OutlierMixin):
+    """Solves linear One-Class SVM using Stochastic Gradient Descent.
+
```
We could add a brief mention to Nystroem here, to extend this class to non-linear kernels.

No need to go in great details, as it is already in the user guide and in both examples, but a brief mention here could still be useful.

TomDLT reviewed (approved):
Only nitpicks.

Strong contribution here, nice work !

albertcthomas committed:
suggestions

Co-authored-by: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

albertcthomas committed:
review

albertcthomas commented:
Thanks a lot @TomDLT for all your reviews and great comments.

ogrisel committed:
Merge branch 'main' into sgd_ocsvm

ogrisel commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1606,3 +1618,442 @@ def _more_tags(self):
+    """Solves linear One-Class SVM using Stochastic Gradient Descent.
+
+    This implementation is meant to be used with a kernel approximation
+    technique to obtain results similar to `sklearn.svm.OneClassSVM` which uses
```
Maybe mention more explicitly:

 ... to be used with in a pipeline with a kernel approximation technique such as `sklearn.kernel_approximation.Nystroem` to obtain results similar to ...

ogrisel commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1606,3 +1618,442 @@ def _more_tags(self):
+        y : array, shape (n_samples,)
+            Labels of the samples.
+        """
+        y = (self.decision_function(X) >= 0).astype(int)
```
Please use `np.int32` to avoid platform specific precisions in datatypes.

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1338,6 +1387,303 @@ def test_loss_function_epsilon(klass):
+    assert_array_equal(clf.coef_, clf2.coef_)
+
+
+@ignore_warnings
```
Can you please add a comment to explain which warning we want to ignore and why is it ok to ignore them in this test?

Please also pass the warning class as the category argument to this decorator, e.g. `@ignore_warnings(category=ConvergenceWarning)`.

ogrisel replied:
Actually isn't there an easy way to avoid triggering the warning in this test in the first place?

albertcthomas replied:
I removed this `ignore_warnings`, apparently no warning is raised.

albertcthomas replied:
Indeed CIs are still green, thanks for spotting this

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1338,6 +1387,303 @@ def test_loss_function_epsilon(klass):
+    assert clf.t_ == t
+    assert_array_almost_equal(y_scores, y_scores2)
+    assert_array_almost_equal(clf.coef_, coef)
+    assert_array_almost_equal(clf.offset_, offset)
```
I think we are standardizing to `assert_allclose` instead of `assert_array_almost_equal` in scikit-learn nowadays.

For this this does not matter too much, but often it's more natural to pass `atol` or `rtol` or combination of both instead of `decimals`.

ogrisel replied:
Similar comment elsewhere in this PR.

albertcthomas replied:
I used `assert_allclose` everywhere I introduced a new `assert_array_almost_equal`. There are remaining ones (involving other SGD estimators) that should preferably be removed in another PR.

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1338,6 +1387,303 @@ def test_loss_function_epsilon(klass):
+    y_pred_sgdocsvm = pipe_sgd.predict(X_test)
+    dec_sgdocsvm = pipe_sgd.decision_function(X_test).reshape(1, -1)
+
+    corrcoef = np.corrcoef(np.concatenate((dec_ocsvm, dec_sgdocsvm)))[0, 1]
+    assert np.mean(y_pred_sgdocsvm == y_pred_ocsvm) >= 0.99
+    assert corrcoef >= 0.9
```
Easier to follow in that order:

```suggestion
    assert np.mean(y_pred_sgdocsvm == y_pred_ocsvm) >= 0.99
    corrcoef = np.corrcoef(np.concatenate((dec_ocsvm, dec_sgdocsvm)))[0, 1]
    assert corrcoef >= 0.9
```

ogrisel commented on sklearn/svm/_classes.py:
```diff
@@ -1334,6 +1334,10 @@ class OneClassSVM(OutlierMixin, BaseLibSVM):
+
+    See also
+    --------
+    sklearn.linear_model.SGDOneClassSVM
```
Maybe add a quick note to point out the favorable linear complexity for datasets with large number of samples.

Here is an example of such notes in a see also section: 

https://github.com/scikit-learn/scikit-learn/blob/95119c13a/sklearn/ensemble/_forest.py#L1727-L1731

ogrisel reviewed (commented):
Hum sorry I broke the tests when resolving the conflicts via github. Let me fix it.

In the mean time here are other comments.

The PR looks very good, and the benchmarks are really impressive. Don't forget to add a what's new entry in `doc/whats_new/v1.0.rst`.

ogrisel committed:
Fix assert_raises => pytest.raises

ogrisel commented on doc/modules/sgd.rst:
```diff
@@ -232,6 +232,57 @@ For regression with a squared loss and a l2 penalty, another variant of
+samples. Note that the complexity of a kernelized One-Class SVM is at best
+quadratic in the number of samples.
+:class:`sklearn.linear_model.SGDOneClassSVM` is thus well suited for datasets
+with a large number of training samples (> 10.000).
```
```suggestion
with a large number of training samples (> 10,000) for which the SGD
variant can be several orders of magnitude faster.
```

ogrisel reviewed (approved):
LGTM once the comments above are taken care of.

One more suggestion:

albertcthomas committed:
Update doc/modules/sgd.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

albertcthomas committed:
Update sklearn/linear_model/tests/test_sgd.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

albertcthomas committed:
avoid insert in example

albertcthomas commented on examples/miscellaneous/plot_anomaly_comparison.py:
```diff
@@ -86,6 +99,14 @@
+clf_sgd = SGDOneClassSVM(nu=outliers_fraction, shuffle=True,
+                         fit_intercept=True, random_state=42, tol=1e-6)
+pipe_sgd = make_pipeline(transform, clf_sgd)
+anomaly_algorithms.insert(2, ("One-Class SVM (SGD)", pipe_sgd))
```
Yes indeed your suggestion is better, thanks

cmarmo commented:
Hi @albertcthomas I believe this is worth a change log entry for 1.0... :)

albertcthomas commented:
> Hi @albertcthomas I believe this is worth a change log entry for 1.0... :)

Yes will do. I've just started working on @ogrisel's review :)

albertcthomas committed:
whats_new entry

albertcthomas committed:
add note

albertcthomas committed:
int32

albertcthomas committed:
explicit ddocstring

albertcthomas committed:
rm useless ignore warnings in tests

albertcthomas committed:
use assert_allclose

albertcthomas commented on benchmarks/bench_online_ocsvm.py:
```diff
@@ -0,0 +1,279 @@
+    # Loading datasets
+    if dataset_name in ['http', 'smtp', 'SA', 'SF']:
+        dataset = fetch_kddcup99(subset=dataset_name, shuffle=False,
+                                 percent10=False, random_state=88)
```
I'll check this. If it's much faster this would definitely be better.

albertcthomas replied:
I tried the two fetchers and for me it seems `fetch_kddcup99` is faster `fetch_openml` Note that to get the full data set (percent10=False) from OpenML you need to set version to 5 (https://www.openml.org/d/42746). I might be missing something.
```python
In [11]: %timeit fetch_openml(name="KDDCup99", return_X_y=True, version=5)
3min 23s ± 975 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [12]: %timeit fetch_kddcup99(percent10=False, return_X_y=True)
28 s ± 114 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

albertcthomas committed:
Merge remote-tracking branch 'upstream/main' into sgd_ocsvm

albertcthomas committed:
remove assert_allclose import

albertcthomas committed:
remaining almost_equal

albertcthomas commented on doc/modules/sgd.rst:
```diff
@@ -232,6 +232,57 @@ For regression with a squared loss and a l2 penalty, another variant of
+solution of a kernelized One-Class SVM, implemented in
+:class:`sklearn.svm.OneClassSVM`, with a linear complexity in the number of
+samples. Note that the complexity of a kernelized One-Class SVM is at best
+quadratic in the number of samples.
```
Hm, with my knowledge of the SMO algorithm and what is written page 26 of the [libsvm paper](https://www.csie.ntu.edu.tw/~cjlin/papers/libsvm.pdf) on its complexity ("Empirically,  it  is  known  that  the  number  of  iterations  may be  higher  than  linear  to  the  number  of  training  data") I am not sure that I can say that the complexity is closer to cubic but you might me more knowledgeable than me on this.

albertcthomas replied:
For the number of support vectors: the fraction of support vectors of the One-Class SVM is greater than `nu` but should converge towards `nu` as the number of training samples becomes large and `nu` is often lower than 0.5 in this case.

ogrisel replied:
Alright let's leave it like this then.

ogrisel commented:
Merged! Thank you very much @albertcthomas. I did not wait for the last open comment of https://github.com/scikit-learn/scikit-learn/pull/10027/files#r598734037 which is minor and can be tackled in a separate PR if you wish.

ogrisel commented:
Thank you again for the very nice contribution!

albertcthomas commented:
Wow thanks a lot!! Thanks again for the first review and all the help @TomDLT :) Thanks for the additional reviews and comments @ogrisel @bthirion @banilo @amueller. Thanks @agramfort for making me work on this :) and thanks @cmarmo for reviving this PR
